sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -444,12 +444,17 @@ object ColumnPruning extends Rule[LogicalPlan] {
       f.copy(child = prunedChild(child, f.references))
     case e @ Expand(_, _, child) if (child.outputSet -- e.references).nonEmpty =>
       e.copy(child = prunedChild(child, e.references))
-    case g: Generate if !g.join && (g.child.outputSet -- g.references).nonEmpty =>
-      g.copy(child = prunedChild(g.child, g.references))
 
-    // Turn off `join` for Generate if no column from it's child is used
-    case p @ Project(_, g: Generate) if g.join && p.references.subsetOf(g.generatedSet) =>
-      p.copy(child = g.copy(join = false))
+    // prune unrequired references
+    case p @ Project(_, g: Generate) =>
+      if (p.references == g.outputSet) {
+        p
+      } else {
+        val requiredAttrs = p.references -- g.producedAttributes ++ g.generator.references
+        val newChild = prunedChild(g.child, requiredAttrs)
+        p.copy(child = g.copy(child = newChild,
+                          unrequiredChildOutput = (g.generator.references -- p.references).toSeq))
```

Comment:
nit: code style
```
val unrequired = (g.generator.references -- p.references).toSeq
p.copy(child = g.copy(child = newChild, unrequiredChildOutput = unrequired)
```
avoid very long statement and create local variables to shorten it.

Reply:
a small improvement to allow us to still hit `case p @ Project(_, child) if sameOutput(child.output, p.output) => child` later
```
case p @ Project(_, g: Generate) if p.references != g.outputSet =>
  ...
```